### PR TITLE
daml-lf: encoder generates dar files

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
With this PR, the daml-lf encoder generate dar file. 
This PR include also generic dar files  for testind runtime.  (Those will be expended and more extensively use in future PRs)   

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/encoder/BUILD.bazel [hunk 1]
```diff
@@ -72,10 +80,82 @@ da_scala_binary(
...
+        ":lf_since_1.3",
+        ":lf_since_1.dev",
+    ],
+    outs = ["test-1.dev.dar"],
+    cmd = "$(location :encoder_binary) $(SRCS) -o $@ -v 1.dev",
+    tools = [":encoder_binary"],
+    visibility = [
+        "//daml-lf:__subpackages__",
+        "//language-support:__subpackages__",
+    ],
```
Comment (S11001001) [inline daml-lf/encoder/BUILD.bazel]: Can you write these `filegroup` and `genrule` calls with list comprehension[s]? See https://github.com/digital-asset/daml/blob/e0e3ce51c7b22ffc78e8f684ed4c0a413d850659/daml-lf/tests/BUILD.bazel for examples.

File: daml-lf/encoder/BUILD.bazel [hunk 2]
```diff
@@ -32,7 +32,12 @@ da_scala_test_suite(
     name = "tests",
     size = "small",
     srcs = glob(["src/test/scala/com/digitalasset/daml/lf/testing/archive/**/*.scala"]),
-    data = [":assembler-test-dalf"],
+    data = [
+        ":testing-dar-1.0",
+        ":testing-dar-1.1",
+        ":testing-dar-1.3",
+        ":testing-dar-1.dev",
+    ],
```
Comment (S11001001) [inline daml-lf/encoder/BUILD.bazel]: The comprehension should also go here.

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala [hunk 3]
```diff
@@ -24,46 +26,118 @@ private[digitalasset] object DamlLfEncoder extends App {
...
+    } catch {
+      case e: EncodeError =>
+        error(s"Encoding error: ${e.message}")
+      case NonFatal(e) =>
+        error(s"error: ${e.getMessage}")
+    }
+
+  private def readSources(files: Seq[String]) = {
+    val builder = StringBuilder.newBuilder
+    files.foreach { file =>
```
Comment (S11001001) [inline daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]: `files.flatMap { file => (??? : String) }(breakOut): String` should typecheck. If you're not a fan of that, go with `map` and `mkString` over this.

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala [hunk 4]
```diff
@@ -24,46 +26,118 @@ private[digitalasset] object DamlLfEncoder extends App {
...
 
+  private case class Arguments(
+      inputFiles: List[String] = List.empty,
+      outputFile: String = "",
+      languageVersion: LanguageVersion = LanguageVersion.default
+  )
+
+  private def parseArgs() = {
+
+    var appArgs = Arguments()
```
Comment (S11001001) [inline daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]:
```suggestion
    @tailrec def go(appArgs: Arguments, i: Int): Arguments = ...
```

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]: fixed

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]: fixed

File: daml-lf/encoder/BUILD.bazel (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/BUILD.bazel]: fixed

File: daml-lf/encoder/BUILD.bazel [hunk 5]
```diff
@@ -10,6 +10,13 @@ load(
     "lf_scalacopts",
 )
 
+lf_target_versions = [
```
Comment (S11001001) [inline daml-lf/encoder/BUILD.bazel]: I suspect you can use the keys of lf_since_deps instead, perhaps moving it up here.

File: daml-lf/encoder/BUILD.bazel [hunk 6]
```diff
@@ -72,10 +82,37 @@ da_scala_binary(
...
+
+lf_since_deps = {
+    "1.0": [],
+    "1.1": ["1.0"],
+    "1.3": ["1.0", "1.1"],
+    "1.dev": ["1.0", "1.1", "1.3"]
+}
+
+[
+    [filegroup(
```
Comment (S11001001) [inline daml-lf/encoder/BUILD.bazel]:
```suggestion
    filegroup(
```

File: daml-lf/encoder/BUILD.bazel [hunk 7]
```diff
@@ -72,10 +82,37 @@ da_scala_binary(
...
+         name = "lf_since_%s" % target,
+         srcs = [":lf_since_%s" %dep for dep in lf_since_deps[target]] +
+           glob(["src/test/lf/since_%s/**/*.lf" % target]),
+    )]
+    for target in lf_target_versions
+]
+
+
+[
+  [genrule(
```
Comment (S11001001) [inline daml-lf/encoder/BUILD.bazel]:
```suggestion
  genrule(
```
Also `bazel run //:buildifier-fix` I'd wager.

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala [hunk 8]
```diff
@@ -24,46 +27,113 @@ private[digitalasset] object DamlLfEncoder extends App {
...
+
+    } catch {
+      case e: EncodeError =>
+        error(s"Encoding error: ${e.message}")
+      case NonFatal(e) =>
+        error(s"error: ${e.getMessage}")
+    }
+
+  private def readSources(files: Seq[String]) =
+    files.flatMap(file => Source.fromFile(Paths.get(file).toFile, "UTF8")).mkString
```
Comment (S11001001) [inline daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]: `flatMap` and `mkString` will make an intermediate `Seq[Char]` all boxed first 😨 
```suggestion
    files.map(file => Source.fromFile(Paths.get(file).toFile, "UTF8")).mkString
```

File: daml-lf/encoder/BUILD.bazel (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/BUILD.bazel]: fixed

File: daml-lf/encoder/BUILD.bazel (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/BUILD.bazel]: fixed

File: daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/src/main/scala/com/digitalasset/daml/lf/testing/archive/DamlLfEncoder.scala]: sorry !

File: daml-lf/encoder/BUILD.bazel (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/encoder/BUILD.bazel]: fixed

**Pull request merged**